tsconvert/tables.py

```diff
@@ -0,0 +1,81 @@
+#
+# MIT License
+#
+# Copyright (c) 2021 Tskit Developers
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+#
+import pandas
+import tskit.util
+
+
+def pandas_to_table_metadata(table, dataframe, index_metadata_property):
+    """
+    Adds metadata to a table, in place, from a pandas dataframe. The dataframe index
+    is matched to the metadata property specified.
+
+    :param AnyTableClass table: The table to modify
+    :param pandas.dataframe dataframe: The dataframe to insert as metadata
+    :param string index_metadata_property: The metadata property to match to the
+        dataframe index.
+    """
+
```

Comment:
Shouldn't we add (or at least give the option to add) a JSON metadata schema here?

Reply:
I think a future enhancement would be to create one from the pandas column specification, I'll add an issue for that. As this function takes a table, I think the caller should use the normal API to add the schema they want. The code as written doesn't need JSON specifically, just any codec that supports dicts.